chore: Managed Netvar documentation (#818)
Unity-Technologies/com.unity.multiplayer.docs: docs/advanced-topics/custom-serialization.md

title: Custom Serialization
---

> When using `RPC`s, `NetworkVariable`'s or any other Netcode for GameObjects (Netcode) related task that requires serialization. The Netcode uses a default serialization pipeline that looks like this:

Reviewer: ```suggestion
When using `RPCs, `NetworkVariable`'s or any other Netcode for GameObjects (Netcode) related task that requires serialization, the Netcode uses a default serialization pipeline that looks like this:
```

Author: I made a commit with the other change and also removed the errant apostrophe from `NetworkVariable`'s.

(Just want to note that the only thing I changed about this line to begin with was removing an errant apostrophe from `RPC`'s. The rest of this copy predated this PR, but it it a valid grammatical change so I went ahead and made it.)

> When using `RPC`s, `NetworkVariable`'s or any other Netcode for GameObjects (Netcode) related task that requires serialization. The Netcode uses a default serialization pipeline that looks like this:

Reviewer 2: Is there a way to make this a full sentence?

> When using `RPC`s, `NetworkVariable`'s or any other Netcode for GameObjects (Netcode) related task that requires serialization.

By default, any type that satisfies the `unmanaged` generic constraint can be automatically serialized as RPC parameters. This includes all basic types (bool, byte, int, float, enum, etc) as well as any structs that contains only these basic types.

> With this flow, you can override **ALL** serialization for **ALL** types, even built in types, and with the API provided, it can even be done with types that you have not defined yourself, those who are behind a 3rd party wall, such as .NET types. However, the way custom serialization is implemented for RPCs and NetworkVariables is slightly different.

Reviewer: ```suggestion
With this flow, you can override **all** serialization for **all** types, even built-in types, and with the API provided, it can even be done with types that you have not defined yourself, those who are behind a 3rd party wall, such as .NET types. However, how custom serialization is implemented for RPCs and NetworkVariables is slightly different.
```

Author: I feel like "the way this is implemented" is more grammatically correct (or at least sounds more natural to me) than "how this is implemented", in this context...

Author: Made a commit that replaced that whole first sentence, because it wasn't really accurate anyway.

    public static void WriteValueSafe(this FastBufferWriter writer, in Url value)
    {

>         writer.WriteValueSafe(value.Value);

Reviewer 2: minor nit-pick.
could we change this:
`in Url value`
To something like this:
`in Url urlValue`

(just trying to avoid `value.Value`)

}
```

Additionally, you can also add extensions for `FastBufferReader.ReadValue()`, `FastBufferWriter.WriteValue()`, and `BufferSerializer<TReaderWriter>.SerializeValuePreChecked()` to provide more optimal implementations for manual serialization using `FastBufferReader.TryBeginRead()`, `FastBufferWriter.TryBeginWrite()`, and `BufferSerializer<TReaderWriter>.PreCheck()`, respectively. However, none of these will be used for serializing RPCs - only `ReadValueSafe` and `WriteValueSafe` are used.

### For NetworkVariable

> `NetworkVariable` goes through a slightly different pipeline than `RPC`s and relies on a different process for determining how to serialize its types. As a result, making a custom type available to the `RPC` pipeline doesn't automatically make it available to the `NetworkVariable` pipeline, and vice-versa. The same method can be used for both, but currently, `NetworkVariable` requires an additional runtime step to make it aware of the methods.

Reviewer: ```suggestion
`NetworkVariable` goes through a slightly different pipeline than `RPCs and relies on a different process for determining how to serialize its types. As a result, making a custom type available to the `RPC` pipeline doesn't automatically make it available to the `NetworkVariable` pipeline, and vice-versa. The same method can be used for both, but currently, `NetworkVariable` requires an additional runtime step to make it aware of the methods.
```

Author: Same comment as above - that's a backtick (inline code formatting), not an apostrophe.

> To add custom serialization support in `NetworkVariable`, follow the steps from the "For RPCs" section to write extension methods for `FastBufferReader` and `FastBufferWriter`; then, somewhere in your application startup (before any `NetworkVariable`s using the affected types will be serialized) add the following:

Reviewer 2: Could you provide a relative link to the "For RPCs" section reference?

Author: How do I provide a link to a heading in the same file?
